candid_diff/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+// Logging:
+#[macro_use]
+extern crate log;
+extern crate env_logger;
+
+use std::rc::Rc;
+
+pub use ::candid::parser::value::IDLField as Field;
+pub use ::candid::parser::value::IDLValue as Value;
+
+pub use ::candid::types::{Field as TypeField, Label, Type};
+pub use ::candid::Nat;
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum VecEdit<R> {
+    /// insert into position, shifting the following (backward)
+    InsertValue(Nat, Value),
+    /// edit value at position, leaving others in place
+    EditValue(Nat, R),
+    /// remove at given position, shifting the following (forward)
+    RemoveValue(Nat),
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct RcVecEdit(pub Rc<VecEdit<RcValueEdit>>);
+
+/// edit a record's fields: edit some fields' values, drop others
+#[derive(Debug, PartialEq, Clone)]
+pub enum RecordEdit<R> {
+    EditValue(Label, R),
+    DropValue(Label),
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum ValueEdit<R> {
+    /// empty change
+    Skip,
+    /// full change: overwrite with new value.
+    Put(Value),
+    /// edit the sub-value (ignore unchanged Opt label).
+    Opt(R),
+    /// edit the vector, in a sequence of vector edits.
+    Vec(Vec<VecEdit<R>>),
+    /// edit the record, field-wise; ignore unchanged fields.
+    Record(Vec<RecordEdit<R>>),
+    /// edit the variant payload (ignore unchanged label).
+    Variant(R),
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct RcValueEdit(pub Rc<ValueEdit<RcValueEdit>>);
```

Comment:
Why do we need `Rc<_>`? Is it possible to have a cycle in the editing structure?

Reply:
Rc's permit sharing, not cycles.  No, there are no cycles.  Yes, it's generally nice to permit sharing.  Otherwise, I could use `Box`.  Either way is fine.